proposals/0383-allow-let-property-wrapper.md

```diff
@@ -0,0 +1,179 @@
+```
+@propertyWrapper
+class WrapperClass<T> : NSObject {
+  var wrappedValue: T
+
+  init(wrappedValue: T) {
+    self.wrappedValue = wrappedValue
+  }
+}
+
+class C {
+  @WrapperClass let value: Int
+
+  init(v: Int) {
+    value = v
+  }
+}
+```
+Here, `WrapperClass` can be made an immutable property wrapper class instance, preventing any future unintentional changes to the property wrapper class type in this context.
+
+SwiftUI property wrappers may also benefit from a let declaration. For example, `@ScaledMetric` in its simplest usage can be written with a `let` instead:
+```
+struct ContentView: View {
+  @ScaledMetric let imageSize = 10
+
+  var body: some View {
+    Image(systemName: "heart.fill")
+    .resizable()
+    .frame(width: imageSize, height: imageSize)
+  }
+}
+```
+
+Similarly, other SwiftUI property wrappers could be `let` declared when they do not require more than a single initialization.
+
+### Property wrappers with `nonmutating set`
+
+Property wrappers that have a `wrappedValue` property with a `nonmutating set` (e.g., SwiftUI's [`@State`](https://developer.apple.com/documentation/swiftui/state/wrappedvalue) and [`@Binding`](https://developer.apple.com/documentation/swiftui/binding/wrappedvalue)) will preserve the reference semantics of the `wrappedValue` implementation even when marked as a `let` declaration. For example:  
+
+```
+@State let weekday: String = "Monday"
+```
+Here, `weekday` is an immutable instance of the `@State` property wrapper, but its `wrappedValue` storage will retain its mutability and reference type traits. This will translate to:
+```
+private let _weekday: State<String> = State<String>(wrappedValue: "Monday")
+var weekday : String {
+  get { 
+    return _weekday.wrappedValue 
+  }
+  nonmutating set {
+    self._weekday.wrappedValue = value
+  }
+  nonmutating _modify { 
+    yield () 
+  }
+}
+var $weekday: Binding<String> {
+  get { return _weekday.projectedValue }
+}
+```
+
+Marking `weekday` as a let declared property will not remove access to its `wrappedValue`'s nonmutating set and the `wrappedValue` can be assigned via the backing property, `_weekday`:
+
+```
+_weekday.wrappedValue = "Tuesday"
+```
+
+However, this does not affect the immutability of `weekday`, which can only be assigned to once like any ordinary let wrapped property. Any attempt to reassign `weekday` will result in an error:
+
+```
+weekday = "Wednesday" // Error: Cannot assign to value: 'weekday' is a 'let' constant
+```
+
+## Source compatibility
+
+This is an additive feature that does not impact source compatibility.
+
+## Effect on ABI stability/API resilience
+
+This is an additive change that has no direct impact that compromises ABI stability.
+
+## Effect on API resilience
+
+This is an additive change that has no impact on API resilience.
```

Comment:
Are there "Alternatives considered" or "Future directions" applicable here? Or does this proposal really complete the picture in terms of what property wrappers can be expected to support for the foreseeable future?

Reply:
> In reviewing review thread comments, did you ever get around to looking into @ktoso's question about `@TaskLocal`?

Yes, I added [this test here](https://github.com/apple/swift/pull/62342/commits/a983c0740eef3beae00ea1c976a655f9fc2c0c68#diff-6808a8c1547f7f88cc7f05b0f122718c7388a69eed21d9a70b9873ca847f920b) to explicitly check for let wrapped TaskLocal.

You've given me quite a bit to rethink here - thank you! I'll work through this in the next day or so.

After some additional discussion about `nonmutating set`, I think I need to move this part of the proposal to Alternatives Considered section instead. I'll rework that this week as well. Requesting a few more days to work on this before we move forward.

Reply:
Sure—thanks for being so responsive. Ping me when you're ready!